### PR TITLE
Adding nOctaves as parameter in LogSpectrum

### PR DESCRIPTION
**Pull request by xaviliz**
The parameter is defined to 7 as default and it has a range from 1 to 9 octaves.

### CODE REVIEW

**Comment by xaviliz** [comment]
It was already tested in MacOS Catalina.

File: src/algorithms/spectral/logspectrum.cpp [hunk 1]
```diff
@@ -24,68 +24,76 @@ using namespace std;
...
-"References:\n"
-"  [1] Mauch, M., & Dixon, S. (2010, August). Approximate Note Transcription\n"
-"  for the Improved Identification of Difficult Chords. In ISMIR (pp. 135-140).\n"
-"  [2] Chordino and NNLS Chroma,\n"
-"  http://www.isophonics.net/nnls-chroma");
-
-void LogSpectrum::configure() {
+const char *LogSpectrum::name = "LogSpectrum";
+const char *LogSpectrum::category = "Spectral";
+const char *LogSpectrum::description = DOC("This algorithm computes spectrum with logarithmically distributed frequency bins. "
```
Comment (dbogdanov) [inline src/algorithms/spectral/logspectrum.cpp:29]: For consistency with the majority of the algorithms, we do not do code indentation for the DOC strings (something we need to review in the future).

File: src/algorithms/spectral/logspectrum.cpp [hunk 2]
```diff
@@ -24,68 +24,76 @@ using namespace std;
...
+void LogSpectrum::configure()
+{
   // Get parameters.
   _frameSize = parameter("frameSize").toInt();
   _sampleRate = parameter("sampleRate").toFloat();
   _rollon = parameter("rollOn").toFloat();
   _nBPS = parameter("binsPerSemitone").toInt();
-  _nOctave = 7;
-  _nNote = _nOctave * 12 * _nBPS + 2 * (_nBPS/2+1); // a core over all octaves, plus some overlap at top and bottom
+  _nOctaves = parameter("nOctaves").toInt();             //7;
```
Comment (dbogdanov) [inline src/algorithms/spectral/logspectrum.cpp:49]: Avoid the hardcoded default value ("7") in comments, as it is redundant and may (theoretically) change.

File: src/algorithms/spectral/logspectrum.cpp [hunk 3]
```diff
@@ -24,68 +24,76 @@ using namespace std;
...
+  _nNote = _nOctaves * 12 * _nBPS + 2 * (_nBPS / 2 + 1); // a core over all octaves, plus some overlap at top and bottom
   initialize();
 }
 
-
-void LogSpectrum::compute() {
-  const vector<Real>& const_spectrum = _spectrum.get();
+void LogSpectrum::compute()
```
Comment (dbogdanov) [inline src/algorithms/spectral/logspectrum.cpp:54]: We tend to use a different style for curly brackets (see https://essentia.upf.edu/coding_guidelines.html#coding-style-guidelines)

File: src/algorithms/spectral/logspectrum.cpp [hunk 4]
```diff
@@ -24,68 +24,76 @@ using namespace std;
...
+{
+  const vector<Real> &const_spectrum = _spectrum.get();
```
Comment (dbogdanov) [inline src/algorithms/spectral/logspectrum.cpp:56]: Same for `&`, have a look at our current style (for consistency with the majority of the code).

File: src/algorithms/spectral/logspectrum.cpp [hunk 5]
```diff
@@ -147,39 +159,44 @@ void LogSpectrum::compute() {
...
   // TODO: rewrite so that everyone understands what is done here.
   // TODO: make this more general, such that it works with all minoctave,
   // maxoctave and whatever _nBPS (or check if it already does)
 
-  int binspersemitone = _nBPS; 
-  int minoctave = 0;  // this must be 0
-  int maxoctave = 7;  // this must be 7
+  int binspersemitone = _nBPS;
+  int minoctave = 0; // this must be 0
+  //int maxoctave = 7;  // this must be 7
```
Comment (dbogdanov) [inline src/algorithms/spectral/logspectrum.cpp:170]: Remove commented legacy code

File: test/src/unittests/spectral/test_logspectrum.py [hunk 6]
```diff
@@ -18,88 +18,287 @@
...
 
 
-
 from essentia_test import *
 import numpy as np
 
-class TestLogSpectrum(TestCase):
 
+class TestLogSpectrum(TestCase):
     def testRegression(self, frameSize=8192 + 1):
```
Comment (dbogdanov) [inline test/src/unittests/spectral/test_logspectrum.py:26]: There's lots of formatting change that break our current (though poorly consistent) style. I would prefer to only keep changes in the content of the tests for this PR.

File: test/src/unittests/spectral/test_logspectrum.py [hunk 7]
```diff
@@ -109,30 +308,43 @@ def testRegression(self, frameSize=8192 + 1):
...
-        self.assertConfigureFails(LogSpectrum(), { 'frameSize': 1 })
-        self.assertConfigureFails(LogSpectrum(), { 'sampleRate': 0 })
-        self.assertConfigureFails(LogSpectrum(), { 'rollOn': -1})
+        self.assertConfigureFails(LogSpectrum(), {"frameSize": 1})
+        self.assertConfigureFails(LogSpectrum(), {"sampleRate": 0})
+        self.assertConfigureFails(LogSpectrum(), {"rollOn": -1})
+        self.assertConfigureFails(LogSpectrum(), {"nOctaves": 0})
+        self.assertConfigureFails(LogSpectrum(), {"nOctaves": 10})
+
+    def testNOctavesParam(self):
```
Comment (dbogdanov) [inline test/src/unittests/spectral/test_logspectrum.py:328]: This tests zero input for different number of octaves, so we can change the name of this test to better represent it. Possibly better to merge with `testZero` to make it cover all considered cases?

File: test/src/unittests/spectral/test_logspectrum.py [hunk 8]
```diff
@@ -18,88 +18,287 @@
...
-                            86.051285 ,   97.34698  ,   89.825485 ,   62.368706 ,
-                            67.440575 ,   55.217617 ,   40.645    ,   69.08143  ,
-                            58.60874  ,   40.766853 ,   52.087135 ,   75.063484 ])
-
-        audio = MonoLoader(filename = join(testdata.audio_dir, 'recorded/vignesh.wav'),
-                    sampleRate = 44100)()
-
-        w = Windowing(type='hann', normalized=False)
-        spectrum = Spectrum()  # FFT() would return the complex FFT, here we just want the magnitude spectrum
+        expected = array(
```
Comment (dbogdanov) [inline test/src/unittests/spectral/test_logspectrum.py:27]: We can re-use the same hardcoded `expected` vector to add more `assertAlmostEqualVector` asserts for different number of octaves.

**Review by dbogdanov — CHANGES_REQUESTED** [review]
Let's keep only changes related to the content, keeping the code style intact for consistency (see comments inline, of course we need review and improve our [code style guidelines](https://essentia.upf.edu/coding_guidelines.html#coding-style-guidelines)). You can either undo those changes or resubmit another clean PR.

File: src/algorithms/spectral/logspectrum.cpp (hunk 1 above)
Comment (xaviliz) [reply src/algorithms/spectral/logspectrum.cpp:29]: oh yes, you are right. I forgot to switch off the formatter and it modified the code style. No problem, I will resubmit another clean PR and take care of the guidelines.

**Comment by dbogdanov** [comment]
@xaviliz Can you clarify the reasons for closing this PR?

**Comment by xaviliz** [comment]
Oh sorry. I can find no reasons, probably it was a mistake. It should be reopen. 
Please, let me revise your suggestions.

**Pull request closed without merging**